### PR TITLE
services/grpclb: use Stopwatch to provide elapsed time for backoffs

### PR DESCRIPTION
**Pull request by voidzcy**
Use a separate Stopwatch for Load Balancer to manage backoff time. Solves #5056

### CODE REVIEW

File: services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerUtil.java [hunk 1]
```diff
@@ -63,7 +64,9 @@ private HealthCheckingLoadBalancerUtil() {
   public static LoadBalancer newHealthCheckingLoadBalancer(Factory factory, Helper helper) {
     HealthCheckingLoadBalancerFactory hcFactory =
         new HealthCheckingLoadBalancerFactory(
-            factory, new ExponentialBackoffPolicy.Provider(), TimeProvider.SYSTEM_TIME_PROVIDER);
+            factory, new ExponentialBackoffPolicy.Provider(), TimeProvider.SYSTEM_TIME_PROVIDER,
+            Stopwatch
+                .createUnstarted());
```
Comment (dapengzhang0) [inline services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerUtil.java]: nit: `TimeProvider.SYSTEM_TIME_PROVIDER,` in a new line, `Stopwatch.createUnstarted()` in a line.

File: services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java [hunk 2]
```diff
@@ -357,23 +361,23 @@ public String toString() {
...
       private boolean callHasResponded;
 
       HcStream() {
-        callCreationNanos = time.currentTimeNanos();
         callServiceName = serviceName;
         call = subchannel.asChannel().newCall(HealthGrpc.getWatchMethod(), CallOptions.DEFAULT);
       }
 
       void start() {
+        stopwatch.reset().start();
```
Comment (dapengzhang0) [inline services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java]: Why not keep the original behavior?

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java [hunk 3]
```diff
@@ -53,6 +54,7 @@
 
   private final Helper helper;
   private final TimeProvider time;
```
Comment (dapengzhang0) [inline grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java]: `TimeProvider time` seems not needed any more.

Comment (voidzcy) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java]: No, you would need it in `GrpcState` to create a `GrpclbClientLoadRecorder`, which needs to read wall clock.

Comment (dapengzhang0) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java]: Sorry I was wrong, it is used by `GrpclbClientLoadRecorder`

File: services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java (hunk 2 above)
Comment (voidzcy) [reply services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java]: That's something I am also wondering. Wouldn't this be a better way, although the only invocation of HC rpc starts immediately after creation.

https://github.com/grpc/grpc-java/blob/9373b2ce10eec4b60bd10709c7b32e892dafd226/services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java#L325-L326

Comment (dapengzhang0) [reply services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java]: There seems a bug. You should not share the stopwatch across the `HealthCheckingLoadBalancerFactory`. Each `HcStream` should use a new stopwatch, and no need to reset when cancel. I would keep the original behavior.

Comment (dapengzhang0) [reply services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java]: You might want to do something similar to https://github.com/grpc/grpc-java/blob/v1.19.0/netty/src/main/java/io/grpc/netty/NettyClientHandler.java#L105

File: services/src/test/java/io/grpc/services/HealthCheckingLoadBalancerFactoryTest.java [hunk 4]
```diff
@@ -178,8 +179,15 @@ public void setup() throws Exception {
...
+    Supplier<Stopwatch> fakeStopwatchSupplier = new Supplier<Stopwatch>() {
+      @Override
+      public Stopwatch get() {
+        return Stopwatch.createUnstarted(clock.getTicker());
+      }
+    };
     hcLbFactory = new HealthCheckingLoadBalancerFactory(
-        origLbFactory, backoffPolicyProvider, clock.getTimeProvider());
+        origLbFactory, backoffPolicyProvider,
+        fakeStopwatchSupplier);
```
Comment (zhangkun83) [inline services/src/test/java/io/grpc/services/HealthCheckingLoadBalancerFactoryTest.java]: You can use `clock.getStopwatchSupplier()` instead.

File: grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java [hunk 5]
```diff
@@ -291,6 +291,7 @@ public Void answer(InvocationOnMock invocation) throws Throwable {
     when(backoffPolicy2.nextBackoffNanos()).thenReturn(10L, 100L);
     when(backoffPolicyProvider.get()).thenReturn(backoffPolicy1, backoffPolicy2);
     balancer = new GrpclbLoadBalancer(helper, subchannelPool, fakeClock.getTimeProvider(),
+        Stopwatch.createUnstarted(fakeClock.getTicker()),
```
Comment (zhangkun83) [inline grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java]: You can use `fakeClock.getStopwatchSupplier().get()` instead.

**Pull request merged**